File: Samples/iOS_WKWebView+Tealium_Objc/iOS_WKWebView+Tealium_Objc/TealiumHelper.m
//
//  TealiumHelper.m
//  TealiumWatchKitCocoaPodsObjcApp
//
//  Created by Jonathan Wong on 2/13/18.
//  Copyright © 2018 Tealium. All rights reserved.
//

#import "TealiumHelper.h"

@implementation TealiumHelper

static TealiumHelper *tealiumHelper = nil;

+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        tealiumHelper = [[self alloc] init];
    });
    
    return tealiumHelper;
}

+ (instancetype)sharedInstanceWithView:(UIView *)view {
    tealiumHelper = [[self alloc] initWithView:view];
    return tealiumHelper;

}

- (instancetype)init {
    if (self = [super init]) {
        TEALConfiguration *configuration = [TEALConfiguration configurationWithAccount:@"tealiummobile" profile:@"demo-wkwebview" environment:@"qa"];
        [configuration setLogLevel:TEALLogLevelDev];
        
        _tealium = [Tealium newInstanceForKey:@"1" configuration:configuration];
    }
    return self;
}

- (instancetype)initWithView:(UIView *)view {
    if (self = [super init]) {
        TEALConfiguration *configuration = [TEALConfiguration configurationWithAccount:@"tealiummobile" profile:@"demo-wkwebview" environment:@"qa"];
        [configuration setLogLevel:TEALLogLevelDev];
        configuration.view = view;
        
        _tealium = [Tealium newInstanceForKey:@"1" configuration:configuration];
    }
    return self;
}

+ (void)trackEvent:(NSString *)event {
    [[Tealium instanceForKey:@"1"] trackEventWithTitle:event dataSources:nil];
}

+ (void)trackView:(NSString *)view {
    [[Tealium instanceForKey:@"1"] trackEventWithTitle:view dataSources:nil];
}

+ (void)startTracking {
    [TealiumHelper sharedInstance];
}

+ (void)startTrackingWithView:(UIView *)view {
    [TealiumHelper sharedInstanceWithView:view];
}

+ (void)stopTracking {
    [Tealium destroyInstanceForKey:@"1"];
}
@end

File: Samples/WatchOS_Sample_Swift/WatchOS_Sample_Swift/TealiumHelper.swift
//
//  TealiumHelper.swift
//  WatchOS_Sample_Swift
//
//  Created by Jonathan Wong on 3/19/18.
//  Copyright © 2018 Tealium. All rights reserved.
//

import Foundation
import TealiumIOSLifecycle

class TealiumHelper: NSObject {
    
    static let TealiumInstanceIdentifier = "TEALIUM_INSTANCE_1"
    
    /**
     Shared singleton instance of TealiumHelper
     */
    static let shared = TealiumHelper()
    
    let tealium: Tealium
    
    private override init() {
        let configuration = TEALConfiguration(account: "tealiummobile", profile: "demo", environment: "dev")
        configuration.logLevel = TEALLogLevel.dev
        tealium = Tealium.newInstance(forKey: TealiumHelper.TealiumInstanceIdentifier, configuration: configuration)
        super.init()
        tealium.setDelegate(self)
    }
    
    /**
     Creates singleton instance of TealiumHelper
    */
    class func startTracking() {
        _ = TealiumHelper.shared
    }
    
    /**
     Triggers an event tracking call
     @param title any String identifier for the event
     @param data an optional Dictionary of additional data for Tealium mapping.
    */
    class func trackEventWithTitle(title: String, dataSources: [AnyHashable: Any]?) {
        TealiumHelper.shared.tealium.trackEvent(withTitle: title, dataSources: dataSources)
    }
    
    /**
     Triggers a view tracking call
     @param title any NSString identifier for the view
     @param data an optional NSDictionary of additional data for Tealium mapping.
    */
    class func trackViewWithTitle(title: String, dataSources: [AnyHashable: Any]?) {
        TealiumHelper.shared.tealium.trackView(withTitle: title, dataSources: dataSources)
    }
    
    /**
     Forwards the request to the Tealium SDK to optionally make a track call
     */
    class func session(_ session: WCSession, didReceiveMessage message: [String: Any], replyHandler: @escaping ([String: Any]) -> Void) {
        // Turn on the Tealium library if not already on
        startTracking()
        
        // REQUIRED: forwards any track view/event calls to Tealium SDK
        Tealium.instance(forKey: TealiumHelper.TealiumInstanceIdentifier)?.session(session, didReceiveMessage: message, replyHandler: replyHandler)
    }
}

extension TealiumHelper: TealiumDelegate {
    
}
